### PR TITLE
Lwm2m block transfer

### PR DESCRIPTION
**Pull request by LukasWoodtli**
This is an attempt to provide CoAP block 1 transfer with lwm2m.
For now, it's only implemented for the registration message.

This change is meant to be used as a basis for discussions.
I'm aware that there is a lot of room for improvement.

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_object.h [hunk 1]
```diff
@@ -484,10 +487,16 @@ struct lwm2m_message {
 	struct lwm2m_obj_path path;
 
 	/** CoAP packet data related to the outgoing message */
-	struct coap_packet cpkt;
+	struct coap_packet cpkt[CONFIG_LWM2M_COAP_MAX_BLOCK_NUM];
 
 	/** Buffer data related outgoing message */
-	uint8_t msg_data[MAX_PACKET_SIZE];
+	uint8_t msg_data[CONFIG_LWM2M_COAP_MAX_BLOCK_NUM][MAX_PACKET_SIZE];
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_object.h:493]: So the idea here is to store the full message within `struct lwm2m_message`, basically multiplying the struct size by tje factor of  `LWM2M_COAP_MAX_BLOCK_NUM`? I don't think this is acceptable, seems like a heavy RAM waste just to cover rare cases when the blockwise transfer is needed.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 2]
```diff
@@ -172,6 +172,45 @@ static void free_block_ctx(struct lwm2m_block_context *ctx)
 	ctx->tkl = 0U;
 }
 
+static int init_message_pending(struct lwm2m_message *msg)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:182]: It makes it harder to review when the actual changes introduced are mixed up with moving code around (like introducing helper function here). This should be done in a separate commits.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 3]
```diff
@@ -1521,5 +1590,22 @@ static int lwm2m_rd_client_init(const struct device *dev)
 
 }
 
-SYS_INIT(lwm2m_rd_client_init, APPLICATION,
-	 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+#ifdef CONFIG_LWM2M_BLOCK_TRANSFER_TESTS
+/* These wrappers are used for tests to get access to
+ * static functions. They need to be declared manually in the
+ * tests before use. */
+void test_lwm2m_rd_client_setup(struct lwm2m_ctx *ctx)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:1597]: I'm really not a fan of introducing test code directly into the implementation. If you need to access static variables/fucntions, just include the `*.c` file in the test file.

File: subsys/net/lib/lwm2m/Kconfig [hunk 4]
```diff
@@ -597,6 +605,14 @@ config LWM2M_GATEWAY_IOT_DEVICE_OBJECTS_MAX_STR_SIZE
 
 endif # LWM2M_GATEWAY_OBJ_SUPPORT
 
+config LWM2M_BLOCK_TRANSFER_TESTS
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:608]: Ditto, if it's really needed to check if the code is executed in test, use `CONFIG_ZTEST` .

**Review by rlubos — COMMENTED** [review]
It'd be really nice to follow https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines even for draft PRs. It's easier to review if individual commits are introducing logically separated changes (like not mixing CoAP and LwM2M modifications). Plus I find it really inconvenient not having some description in the commit message, it makes it even trickier to review w/o some background information about the general idea behind the change.

Regarding the changes intorduced in this PR, I think that multiplying the `struct lwm2m_message` size by the number of supported blocks is a serious no-go for this approach.

**Comment by LukasWoodtli** [comment]
@rlubos Thank you for your input!

Sorry for the big PR. I'll do smaller commits in future (draft) PR's.

> It's easier to review if individual commits are introducing logically separated changes (like not mixing CoAP and LwM2M modifications).

That's true. But it's really hard to not mix the lwm2m and the CoAP changes in this codebase. Block-transfer is a CoAP feature, so it should not be necessary to chage any lwm2m related code. But a lot of CoAP related things are done in the lwm2m code. I can understand that this is done for optimization reasons. But it doesn't make the code very maintainable and extensible.


> I think that multiplying the `struct lwm2m_message` size by the number of supported blocks is a serious no-go for this approach.

This implies to have some kind of dynamic memory allocation. I'll try another approach with memory slabs (I suppose that's the right memory allocation structure to use in this case).


> I'm really not a fan of introducing test code directly into the implementation. If you need to access static variables/fucntions, just include the *.c file in the test file.

I'm not a fan of including *.c files. This is not how the use of *.c files (translation units) were designed. And it can lead to strange (build) errors. But I'm ok to do that if that's the way it is done in Zephyr. (btw. I'm also not a fan of introducing test code directly into the implementation. I prefer testable implementation design.)


Thank you again for the input. I'm new to the Zephyr project and to CoAP/lwm2m, so my approach might not be very good (yet). But it's a good opportunity to learn and improve.

**Comment by rlubos** [comment]
> > I think that multiplying the `struct lwm2m_message` size by the number of supported blocks is a serious no-go for this approach.
> 
> This implies to have some kind of dynamic memory allocation. I'll try another approach with memory slabs (I suppose that's the right memory allocation structure to use in this case).

Not necessarily. One idea discussed for the block transfer support was to have one or more block transfer contexts, which would contain a large buffer, where the complete message could be encoded (something like the `payload` buffer you've introduced in the RD client). Then, based on the block transfer context, a fragment of this buffer could be encoded into a **single** `lwm2m_message` used to transmit individual block. But, at a time, only a single message structure would be in use, as it can be released on receving an ACK on individual block and then reused for the next one.

I don't think we've considered a dynamic allocation for block transfer specifically, but I recall that an idea for using a dedicated mem heap to allocate the `lwm2m_message` payload instead of using fixed-sized buffers was also considered. Even now, for smaller messages a lot of RAM is wasted with fixed-sized buffer approach.

**Comment by LukasWoodtli** [comment]
> > > I think that multiplying the `struct lwm2m_message` size by the number of supported blocks is a serious no-go for this approach.
> > 
> > 
> > This implies to have some kind of dynamic memory allocation. I'll try another approach with memory slabs (I suppose that's the right memory allocation structure to use in this case).
> 
> Not necessarily. One idea discussed for the block transfer support was to have one or more block transfer contexts, which would contain a large buffer, where the complete message could be encoded (something like the `payload` buffer you've introduced in the RD client). Then, based on the block transfer context, a fragment of this buffer could be encoded into a **single** `lwm2m_message` used to transmit individual block. But, at a time, only a single message structure would be in use, as it can be released on receving an ACK on individual block and then reused for the next one.
> 
> I don't think we've considered a dynamic allocation for block transfer specifically, but I recall that an idea for using a dedicated mem heap to allocate the `lwm2m_message` payload instead of using fixed-sized buffers was also considered. Even now, for smaller messages a lot of RAM is wasted with fixed-sized buffer approach.


That sounds reasonable. But I'm not sure how to decide if a block context buffer would be needed and when not. You know the size of the payload only after it was serialized into a buffer. So you would basically always write into a block context and then if the message would be small enough, you would copy the payload into a "regular" message?

I had recently a slightly different idea (I'm focusing on sending messages with blocks here, but it would be similar for receiving a message in blocks):
- each message has a fixed (relatively small) buffer for the header and options (metadata). This would be reused for each block of the same message with some modifications like the block header.
- each message contains pointers to blocks of memory, each of the size of a CoAP block
- when writing the header and options is finished, and a payload needs to be appended a new block is requested, and the payload is written into that buffer
- when the buffer is full, a new block is requested and more of the payload is written to it. This is repeated until the full payload is written (there might be a limit for number of blocks per message).
- when sending a block over the socket, a message is constructed by combining the metadata (which would need some adjustments) with the current payload block. Receiving a "Continue" would free the current block data and process the next block.

With this approach, each message uses a minimal amount of memory (there might be some wasted memory as the metadata buffer might not be fully used, and the last block will probably not be filled completely). It could possibly also be implemented transparent to the code that serializes the messages. This would require no (or just minimal) change in code outside the CoAP module.
Instead of implementing a block context "by hand" we can reuse functionality provided by Zephyr (e.g. slabs).

Regardless of which implementation we choose, I would suggest using a separate buffer for the metadata (header and options). This makes it easier to reuse it for each block. We will also need to be able to change some data there (i.e. adding/changing the block2 option and adjusting the message ID).

Btw. I see something like a block context also as a kind of dynamic memory allocation, as it is a shared region of memory that can be shared and needs to be managed.

**Comment by LukasWoodtli** [comment]
An additional note:
We had an internal discussion and came to the conclusion that allowing block-transfer could be a configuration option. And that the case where no block transfer would be needed should require only minimal code changes.

**Comment by rlubos** [comment]
> That sounds reasonable. But I'm not sure how to decide if a block context buffer would be needed and when not. You know the size of the payload only after it was serialized into a buffer. So you would basically always write into a block context and then if the message would be small enough, you would copy the payload into a "regular" message?

I agree that's a bit problematic. Ideally, the encoders should enable to do kind of "dry-run" to allow to determine the size needed, but given that JSON or CBOR encoders use 3rd party libraries, that do not support such functionality, it's rather not an option.

Generally, I see 2 viable options here:
* the variation of the one you propose, i. e. encode the payload into some large buffer first, and then based on the size either send it in a single message or initiate a block transfer,
* or (perhaps easier for the initial implementation), use a fallback mechanism, i. e. if the encoding into a regular message buffer fails, fallback to the block transfer. It might not be the best option though if the block transfer is heavily used in the system.

> I had recently a slightly different idea (I'm focusing on sending messages with blocks here, but it would be similar for receiving a message in blocks):

Would be good if you could add this option to https://github.com/zephyrproject-rtos/zephyr/issues/47598, I don't think we've considered this approach earlier. To me it seems like combination of the two approaches proposed in there (serializing whole/on-the-fly):
* It has a benefit over the on-the-fly approach, of not having to worry about the resource value changes during the transfer,
* but would require changes in the encoders/decoders to allow to work properly in scenarios where encoding takes place over multiple buffers

Keeping CoAP headers separate from header might sound nice, but given that we use socket API, it'd have an overhead of copying the header/payload into a serialized buffer to transfer for each message, so that's another thing to consider.

> We had an internal discussion and came to the conclusion that allowing block-transfer could be a configuration option. And that the case where no block transfer would be needed should require only minimal code changes.

I totally agree, a generic block transfer support should be a configurable option.

**Comment by LukasWoodtli** [comment]
I added a new PR (somewhat different approach) here #54761.
